### PR TITLE
ctfe: regularly get-sth to keep metrics fresh

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: trillian/ctfe/ct_server/main.go [hunk 1]
```diff
@@ -44,6 +45,7 @@ var (
 	metricsEndpoint    = flag.String("metrics_endpoint", "localhost:6963", "Endpoint for serving metrics; if left empty, metrics will be visible on --http_endpoint")
 	rpcBackendFlag     = flag.String("log_rpc_server", "localhost:8090", "Backend specification; comma-separated list or etcd service name (if --etcd_servers specified)")
 	rpcDeadlineFlag    = flag.Duration("rpc_deadline", time.Second*10, "Deadline for backend RPC requests")
+	getSTHInterval     = flag.Duration("get_sth_interval", time.Second*30, "Interval between internal get-sth operations (0 to disable)")
```
Comment (Martin2112) [inline trillian/ctfe/ct_server/main.go]: 30 seconds possibly too short if we have a large number of logs eventually? Will do for now.

File: trillian/ctfe/ct_server/main.go [hunk 2]
```diff
@@ -140,6 +142,23 @@ func main() {
 		http.Handle("/metrics", promhttp.Handler())
 	}
 
+	if *getSTHInterval > 0 {
+		// Regularly update the internal STH for each log so our metrics stay up-to-date with any tree head
+		// changes that are not triggered by us.
+		for _, c := range cfg {
```
Comment (Martin2112) [inline trillian/ctfe/ct_server/main.go]: I'm wondering whether this should be a single background operation that cycles round all the logs. I don't have a firm feeling that it's overall better though.

File: trillian/ctfe/ct_server/main.go [hunk 3]
```diff
@@ -140,6 +142,23 @@ func main() {
...
+			ticker := time.NewTicker(*getSTHInterval)
+			go func(c *configpb.LogConfig) {
+				glog.Infof("start internal get-sth operations on log %v (%d)", c.Prefix, c.LogId)
+				for t := range ticker.C {
+					glog.V(1).Infof("tick at %v: force internal get-sth for log %v (%d)", t, c.Prefix, c.LogId)
+					if _, err := ctfe.GetTreeHead(ctx, client, c.LogId, c.Prefix); err != nil {
+						glog.Warningf("failed to retrieve tree head for log %v (%d): %v", c.Prefix, c.LogId, err)
```
Comment (Martin2112) [inline trillian/ctfe/ct_server/main.go]: Does this need its own metric or is it covered elsewhere?

File: trillian/ctfe/ct_server/main.go (hunk 2 above)
Comment (daviddrysdale) [inline trillian/ctfe/ct_server/main.go]: In Go-land, goroutines are supposed to be cheap so I think one per-log makes sense.

File: trillian/ctfe/ct_server/main.go (hunk 1 above)
Comment (daviddrysdale) [inline trillian/ctfe/ct_server/main.go]: Upped to 180 seconds.

File: trillian/ctfe/ct_server/main.go (hunk 3 above)
Comment (daviddrysdale) [inline trillian/ctfe/ct_server/main.go]: Given that the main point here is to keep another metric updated, it didn't seem worth tracking -- but I can add one if you disagree.

Comment (Martin2112) [inline trillian/ctfe/ct_server/main.go]: If we fail to update STHs though does that become visible without external traffic. If so then I think we're good.

**Pull request merged**